### PR TITLE
[OGUI-1892] Disable level buttons for shifter users on page load and after

### PR DESCRIPTION
**Pull request by graduta**
#### I have JIRA issue created
- [x] branch and/or PR name(s) includes JIRA ID
- [x] issue has "Fix version" assigned
- [x] issue "Status" is set to "In review"
- [x] PR labels are selected
- [x] FLP integration tests were ran successful

PR which:
* if user authenticated has a role of  "shifter" but not "admin":
    *  the UI will disable the buttons changing the "level" of messages filtering system
    * even if the URL contains a higher level, this level will be brought back to level 1 (operations) 
* adds tests to validated newly added changes
* adds const/enums for roles and infologger levels
* extracts button into a reusable component and defines objects for existing filters

### CODE REVIEW

File: InfoLogger/public/logFilter/commandFilters.js [hunk 1]
```diff
@@ -14,92 +14,97 @@
...
-  },
-  title: title,
-}, label);
+  h(
+    '.btn-group',
+    filterLevelsAllowed.map(({ label, index, available }) => _selectableButtonComponent(
+      label,
+      {
+        id: `level-${index}`,
+        title: available ? `Filter level ≤ ${index}` : `You don't have access to level ${label}`,
```
Comment (isaachilly) [inline InfoLogger/public/logFilter/commandFilters.js]: I think the tooltip will show `Filter level ≤ null` for Trace as it doesn't have an index.

File: InfoLogger/public/logFilter/commandFilters.js [hunk 2]
```diff
@@ -14,92 +14,97 @@
...
- * @param {string} label - button's label
- * @param {string} title - button's title on mouse over
- * @param {string} value - a char to represent severity: W E F or I, can be many with spaces like 'W E'
- * @returns {vnode} - the button to toggle severity
- */
-const buttonSeverity = (model, label, title, value) => h('button.btn', {
-  className: model.log.filter.criterias.severity.in.includes(value) ? 'active' : '',
-  onclick: (e) => {
-    model.log.setCriteria('severity', 'in', value);
-    e.target.blur(); // remove focus so user can 'enter' without actually toggle again the button
```
Comment (isaachilly) [inline InfoLogger/public/logFilter/commandFilters.js:66]: Was `e.target.blur()` intentionally removed?

File: InfoLogger/public/logFilter/commandFilters.js [hunk 3]
```diff
@@ -14,92 +14,97 @@
...
+        },
+      )),
+  ),
+
+  _selectableButtonComponent(
+    'Reset filters',
+    {
+      title: 'Reset date, time, matches, excludes, log levels',
+      isActive: false,
+      onclick: () => logModel.filter.resetCriteria(),
```
Comment (isaachilly) [inline InfoLogger/public/logFilter/commandFilters.js:87]: Reset filters will allow resetting to a default, more permissive criteria set.

File: InfoLogger/public/Model.js [hunk 4]
```diff
@@ -311,12 +311,16 @@ export default class Model extends Observable {
...
    * Delegates sub-model actions depending new location of the page
+   * If user is shifter but not admin, set Ops as maximum level for filtering
    */
   handleLocationChange() {
     const { params } = this.router;
     if (params) {
       this.parseLocation(params);
     }
+    if (hasShifterButNoAdminRole(this.session.access)) {
+      this.log.filter.setCriteria('level', 'max', 1);
```
Comment (isaachilly) [inline InfoLogger/public/Model.js]: Could use infologger-level.const.js?

File: InfoLogger/test/public/shifter-based-actions-mocha.js [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+      const location = await page.evaluate(() => window.location);
+      const search = decodeURIComponent(location.search);
+      assert.strictEqual(search, '?q={"severity":{"in":"I W E F"},"level":{"max":1}}');
+
+      const criterias = await page.evaluate(() => window.model.log.filter.criterias);
+      assert.strictEqual(criterias.level.max, 1);
+    },
+  );
+
+  it('should disable buttons for level filter if user is shifter but not admin', async () => {
```
Comment (isaachilly) [inline InfoLogger/test/public/shifter-based-actions-mocha.js:61]: A couple of edge cases should there be a test for checking what happens if user has both admin and shifter roles?

What should happen if user doesn't have shifter or admin roles?

File: InfoLogger/public/common/utils.js [hunk 6]
```diff
@@ -60,3 +63,30 @@ export function setBrowserTabTitle(title = undefined) {
...
+  return access.includes(Role.SHIFTER) && !access.includes(Role.ADMIN);
+}
+
+/**
+ * Method to return filter levels allowed for filtering based on current user role email groups affiliation
+ * * Shifters are only allowed to filter by Ops level
+ * @param {string[]} access - array of user roles email groups affiliation
+ * @returns {{label: string, index:number}[]} - filter levels allowed for filtering
+ */
+export function getFilterLevelsAllowed(access = []) {
```
Comment (isaachilly) [inline InfoLogger/public/common/utils.js:82]: I comment here, but the change would be elsewhere The access is not checked server-side? If the user changes it locally they can still query things they aren't allowed to. Are we okay with this level of security?

File: InfoLogger/public/Model.js [hunk 7]
```diff
@@ -311,12 +311,16 @@ export default class Model extends Observable {
 
   /**
    * Delegates sub-model actions depending new location of the page
+   * If user is shifter but not admin, set Ops as maximum level for filtering
    */
   handleLocationChange() {
```
Comment (isaachilly) [inline InfoLogger/public/Model.js:319]: Overwriting on each navigation call could be unnecessary and setCriteria could be a better more central place for it to be checked and set to a default.

**Review by isaachilly — COMMENTED** [review]
Please see my thoughts and comments.

File: InfoLogger/public/common/utils.js (hunk 6 above)
Comment (graduta) [reply InfoLogger/public/common/utils.js:82]: This is not a security problem because our interest is not to hide these logs for shifter levels, but simply to help them better navigate the logs. 

If they modify it themselves locally via scripts or curl commands, it means they have a reason for looking at a larger set of logs

File: InfoLogger/public/logFilter/commandFilters.js (hunk 1 above)
Comment (graduta) [reply InfoLogger/public/logFilter/commandFilters.js]: Indeed, updated

File: InfoLogger/public/logFilter/commandFilters.js (hunk 2 above)
Comment (graduta) [reply InfoLogger/public/logFilter/commandFilters.js:66]: From my point of view this was not needed from the start, because the key handler in Model takes care of it. The button will not impact the "enter" for quering the logs

File: InfoLogger/public/logFilter/commandFilters.js (hunk 3 above)
Comment (graduta) [reply InfoLogger/public/logFilter/commandFilters.js:87]: Indeed, good catch! 
I have updated it so that we apply the last available/allowed level

File: InfoLogger/public/Model.js (hunk 4 above)
Comment (graduta) [reply InfoLogger/public/Model.js]: Yes, I should have done that from start!

File: InfoLogger/public/Model.js (hunk 7 above)
Comment (graduta) [reply InfoLogger/public/Model.js:319]: My reasoning was that I want to keep LogFilter decoupled from the logic of roles.
In order to do so and apply your suggestion, I introduced constraing in the filter which gets taken into account when setting criteria.
This in turn will also help with the `resetCriteria` logic

File: InfoLogger/test/public/shifter-based-actions-mocha.js (hunk 5 above)
Comment (graduta) [reply InfoLogger/test/public/shifter-based-actions-mocha.js:61]: Added

**Review by isaachilly — APPROVED** [review]

**Comment by graduta** [comment]
Closing this PR for now, details in the Jira ticket
